src/windows/network.rs

```diff
@@ -246,4 +287,137 @@ impl NetworkExt for NetworkData {
+            return None;
+        }
+        unsafe {
+            let adapter = self.adapter;
+            if let Ok(interface_name) = u16_ptr_to_string((*adapter).FriendlyName).into_string() {
+                if self.unicast_address.is_null() {
+                    // if we found that this interface has not been visited yet,
+                    // set unicast address,
+                    self.unicast_address = (*adapter).FirstUnicastAddress;
+                    // and return the MAC address instead
+                    let [mac @ .., _, _] = (*adapter).PhysicalAddress;
+                    Some((
+                        interface_name,
+                        InterfaceAddress::MAC(MacAddr::from(mac)),
+                    ))
+                } else {
+                    // otherwise, generate an IP adddress
+                    let address = self.unicast_address;
+                    self.unicast_address = (*address).Next;
+                    if self.unicast_address.is_null() {
+                        // if we have visited all unicast addresses, move to next adapter
+                        self.adapter = (*adapter).Next;
+                    }
+                    // FIXME: should we perform a null check on (*address).Address.lpSockaddr
+                    let sock_addr = (*address).Address.lpSockaddr;
+
+                    match (*sock_addr).sa_family as _ {
+                        AF_INET => {
+                            let sock_addr = sock_addr as *const SOCKADDR_IN;
+                            let sock_addr = (*sock_addr).sin_addr.S_un.S_addr();
+                            let mut subnet_mask = 0 as ULONG;
+                            // only avaialbe on vista and later
+                            // https://learn.microsoft.com/zh-cn/windows/win32/api/netioapi/nf-netioapi-convertlengthtoipv4mask
+                            ConvertLengthToIpv4Mask(
+                                (*address).OnLinkPrefixLength as _,
+                                &mut subnet_mask as _,
+                            );
+
+                            Some((
+                                interface_name,
+                                InterfaceAddress::IPv4(
+                                    Ipv4Addr::from(htonl(*sock_addr)),
+                                    Ipv4Addr::from(htonl(subnet_mask)),
+                                ),
+                            ))
+                        }
+                        _ => Some((interface_name, InterfaceAddress::NotImplemented)),
+                    }
+                }
+            } else {
+                // Not sure whether error can occur when parsing adapter name.
+                // If we met an error, move to the next adapter
+                self.adapter = (*adapter).Next;
+                self.next()
+            }
+        }
+    }
+}
+
+impl Drop for InterfaceAddressIterator {
+    fn drop(&mut self) {
+        unsafe {
+            libc::malloc(self.buf as _);
```

Comment:
You malloc a pointer in the drop. Interesting move :rofl:

Reply:
My fault. I was going to call the free instead ...